### PR TITLE
Refactored wayland_client.c to support the virtual platform

### PR DESCRIPTION
**Pull request by mattkae**
## What's new?
- Fixes https://github.com/MirServer/mir/issues/3102
- Rendering in the same manner as ubuntu-frame does it so that everything works with the virtual platform

## To test:
Run this and see that we don't block any longer:
```sh
WAYLAND_DISPLAY=wayland-99 MIR_SERVER_PLATFORM_DISPLAY_LIBS=mir:virtual MIR_SERVER_VIRTUAL_OUTPUT=1280x1024 ./bin/mir_demo_server --test-timeout=3 --test-client ./bin/mir_demo_client_wayland
```

### CODE REVIEW

File: examples/client/wayland_client.c [hunk 1]
```diff
@@ -482,6 +488,70 @@ static struct xdg_toplevel_listener const shell_toplevel_listener =
     handle_xdg_toplevel__capabilities,
 };
 
+void run(struct wl_display* display, int* shutdown_signal, int* draw_signal, draw_context* ctx)
```
Comment (AlanGriffiths) [inline examples/client/wayland_client.c:491]: As [g_]shutdown_signal and [g_]draw_signal are global, why pass them as parameters?

**Review by AlanGriffiths — APPROVED** [review]
Yes.

Feel free to deal with the nits or not before landing

File: examples/client/wayland_client.c (hunk 1 above)
Comment (mattkae) [reply examples/client/wayland_client.c:491]: This was my minor effort to make it less global-ly. But that might be a losing battle, I'm okay with using globals

File: examples/client/wayland_client.c [hunk 2]
```diff
@@ -218,11 +220,18 @@ find_free_buffer(draw_context* ctx)
     return NULL;
 }
 
-static void draw_new_stuff(void* data, struct wl_callback* callback, uint32_t time);
+static int g_draw_signal;
```
[the remark this answers is not in the document]
Comment (mattkae) [reply examples/client/wayland_client.c:223]: Eh I didn't want to shadow the locals, but if I don't need them, then I'll get rid of them

**Pull request merged**